ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,772 @@
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    sync::{Arc, Mutex, Weak},
+};
+
+use async_trait::async_trait;
+use bitvec::{array::BitArray, prelude::Lsb0, slice::BitSlice};
+use futures::{future, Future, Stream};
+use futures_util::{stream::iter, StreamExt};
+
+use crate::{
+    error::Error,
+    ff::Field,
+    protocol::{
+        context::{
+            dzkp_malicious::DZKPUpgraded as MaliciousDZKPUpgraded,
+            dzkp_semi_honest::DZKPUpgraded as SemiHonestDZKPUpgraded, Base, Context,
+            MaliciousContext, SemiHonestContext, UpgradableContext,
+        },
+        step::Gate,
+        RecordId,
+    },
+    seq_join::{seq_join, SeqJoin},
+    sharding::ShardBinding,
+};
+
+/// `UnverifiedValues` are intermediate values that occur during a multiplication.
+/// These values need to be verified since there might have been malicious behavior.
+/// For a multiplication, let `(x1, x2), (x2, x3), (x3,x1)`, `(y1, y2), (y2, y3), (y3,y1)` be the shares of `x` and `y`
+/// and let the goal of the multiplication be to compute `x*y`.
+/// `(pi, p(i+1)` is the randomness from `PRSS` used for the multiplication by helper `i`.
+/// `z(i+1)` is the result of the multiplication received from the helper on the right.
+/// We do not need to store `zi` since it can be computed from the other stored values.
+#[derive(Clone, Debug)]
+struct UnverifiedValues {
```

Comment:
I wonder if it would be worth adding `begin: u8` and `end: u8` here so that we can track the validity within these arrays. Or else in some wrapper type that would take the place of `Option<UnverifiedValues>`. The space overhead is the same or negligibly more than the `Option`. We don't even necessarily need to do all the accounting right away -- to start with we could just implement `is_some` and `is_none` methods in terms of `begin != end`. This would at least give us a framework in case we start having problems with inconsistent batches.

Reply:
This only seems to work when the missing element is at the end or beginning of the array. When I am missing one in between, begin, end would not be helpful. Is there another reason to use begin and end other than checking consistency of the batches?

Reply:
It would let us at least get an exact size for the batch, which might make some of the other calculations more accurate.